SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,615 @@
+
+FInterestManager::FInterestManager(InterestFactory& InInterestF, FSpatialPositionStorage& InPositions,
+								   FAlwaysRelevantStorage& InAlwaysRelevant, FServerAlwaysRelevantStorage& InServerAlwaysRelevant)
+	: InterestF(InInterestF)
+	, Positions(InPositions)
+	, AlwaysRelevant(InAlwaysRelevant)
+	, ServerAlwaysRelevant(InServerAlwaysRelevant)
+{
+	SlotMap.Reserve(1 << 16);
+	Entities.Reserve(1 << 16);
+	EntityPosition.Reserve(1 << 16);
+	EntityFlags.Reserve(1 << 16);
+	Visibility.Reserve(1 << 16);
+	ActiveTimestamp.Reserve(1 << 16);
+}
+
+void FInterestManager::Advance(const TSet<Worker_EntityId_Key>& DeletedEntities)
+{
+	for (Worker_EntityId Entity : DeletedEntities)
+	{
+		if (int32* SlotPtr = SlotMap.Find(Entity))
+		{
+			Remove(*SlotPtr);
+			if (Broadphase)
+			{
+				Broadphase->Remove(*SlotPtr);
+			}
+			SlotMap.Remove(Entity);
+		}
+	}
+
+	for (Worker_EntityId Entity : Positions.GetModifiedEntities())
+	{
+		int32 Slot;
+		int32* SlotPtr = SlotMap.Find(Entity);
+		const FVector& Pos = Positions.GetPositions()[Entity];
+		if (SlotPtr != nullptr)
+		{
+			Slot = *SlotPtr;
+			if (Broadphase)
+			{
+				Broadphase->Update(Slot, Pos);
+			}
+		}
+		else
+		{
+			// Assume these flags are available immediately and never change.
+			uint32 Flags = 0;
+			if (AlwaysRelevant.GetObjects().Contains(Entity))
+			{
+				Flags |= AlwaysRelevantFlag;
+			}
+			if (ServerAlwaysRelevant.GetObjects().Contains(Entity))
+			{
+				Flags |= ServerAlwaysRelevantFlag;
+			}
+			Slot = Allocate();
+			SlotMap.Add(Entity, Slot);
+			Entities[Slot] = Entity;
+			EntityFlags[Slot] = Flags;
+			if (Broadphase)
+			{
+				Broadphase->Add(Slot, Pos);
+			}
+		}
+
+		EntityPosition[Slot] = FVector2D(Pos.X, Pos.Y);
+		EntityPositionX[Slot] = Pos.X;
+		EntityPositionY[Slot] = Pos.Y;
+		ActiveTimestamp[Slot] = FPlatformTime::Cycles64();
+
+		if (Slot >= ActiveEntities)
+		{
+			SwapEntries(ActiveEntities, Slot);
+			++ActiveEntities;
+		}
+	}
+}
+
+int32 FInterestManager::Allocate()
+{
+	Entities.AddDefaulted();
+	EntityPosition.AddDefaulted();
+	EntityPositionX.AddDefaulted();
+	EntityPositionY.AddDefaulted();
+	EntityFlags.AddDefaulted();
+	Visibility.AddDefaulted();
+	ActiveTimestamp.AddDefaulted();
+
+	++ActiveEntities;
+
+	int32 NumEntities = Entities.Num();
+	if (ActiveEntities != NumEntities)
+	{
+		MoveEntry(ActiveEntities - 1, NumEntities - 1);
+	}
+	return ActiveEntities - 1;
+}
+
+void FInterestManager::Remove(int32 Slot)
+{
+	if (Slot < ActiveEntities)
+	{
+		SwapEntries(ActiveEntities - 1, Slot);
+		Slot = ActiveEntities - 1;
+		--ActiveEntities;
+	}
+	if (Slot < Entities.Num() - 1)
+	{
+		SwapEntries(Slot, Entities.Num() - 1);
+	}
+	Entities.Pop(false);
+	EntityPosition.Pop(false);
+	EntityPositionX.Pop(false);
+	EntityPositionY.Pop(false);
+	EntityFlags.Pop(false);
+	Visibility.Pop(false);
+}
+
+void FInterestManager::MoveEntry(int32 SlotSrc, int32 SlotDest)
+{
+	Worker_EntityId EntitySrc = Entities[SlotSrc];
```

Comment:
are we interested in any kind of bounds checking / warning here?

Reply:
I don't think so, currently the slot is only retrieved/used internally